File: src/main/java/com/sem/keeper/web/AjaxController.java
package com.sem.keeper.web;

import com.sem.keeper.entity.DeviceEntity;
import com.sem.keeper.entity.LoanEntity;
import com.sem.keeper.entity.LoanRequestEntity;
import com.sem.keeper.entity.UserEntity;
import com.sem.keeper.repo.DeviceRepository;
import com.sem.keeper.repo.LoanRepository;
import com.sem.keeper.repo.LoanRequestRepository;
import com.sem.keeper.repo.UserRepository;
import com.sem.keeper.service.LoanService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.web.bind.annotation.*;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import javax.servlet.http.HttpSession;
import java.util.Collection;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;


@Service
@RestController
@RequestMapping("/ajax")
public class AjaxController {

    @Autowired
    DeviceRepository devices;

    @Autowired
    LoanRepository loanRepository;

    @Autowired
    LoanService loanService;

    @Autowired
    UserRepository userRepository;

    @Autowired
    LoanRequestRepository loanRequestRepository;

    @GetMapping("/myReqs")
    public Iterable<LoanRequestEntity> myReqs(HttpSession session){
        UserEntity user = (UserEntity) session.getAttribute("user");
        return loanRequestRepository.findByElvinne(user);
    }

    @GetMapping("/devices")
    public Iterable<DeviceEntity> devicePag(@RequestParam int page, @RequestParam int pageSize){
        Pageable pageable = PageRequest.of(page, pageSize);
        return devices.findAll(pageable);
    }

    @GetMapping(value = {"/devicesearch","/devicesearch/{pageNumber}","/devicesearch/{pageNumber}/{pageSize}"})
    public Page<DeviceEntity> searchDevice(@RequestParam String term,
                                           @PathVariable Optional<Integer> pageNumber,
                                           @PathVariable Optional<Integer> pageSize){
        Integer page = pageNumber.orElse(0);
        var pageable = PageRequest.of(page,pageSize.orElse(10));
        return devices.findByNameContainingIgnoreCase(term, pageable);
    }

    @GetMapping("/usersearch/{pageNumber}/{pageSize}")
    public Page<UserEntity> searchUser(@RequestParam String term,
                                           @PathVariable Optional<Integer> pageNumber,
                                           @PathVariable Optional<Integer> pageSize){
        Integer page = pageNumber.orElse(0);
        var pageable = PageRequest.of(page,pageSize.orElse(10));
        return userRepository.findByFullNameContainingIgnoreCase(term, pageable);
    }

    @GetMapping(value = {"/loanlist","/loanlist/{pageNumber}/{pageSize}"})
    public Page<LoanEntity> loanlist(HttpSession session,
                                     @RequestParam Optional<String> filter,
                                     @PathVariable Optional<Integer> pageNumber,
                                     @PathVariable Optional<Integer> pageSize){
        Pageable pageable = PageRequest.of(pageNumber.orElse(0),pageSize.orElse(10));
        if (filter.isEmpty() || filter.get().equals("all")) {
            return loanRepository.findAll(pageable);
        } else {
            switch (filter.get()){
                case "mine":{
                    UserEntity user = (UserEntity) session.getAttribute("user");
                    return loanRepository.findByElvitte(user, pageable);
                }
                case "out":{
                    return loanRepository.findByVisszavetteIsNullOrderByTakeDate(pageable);
                }
                default:{
                    return loanRepository.findAll(pageable);
                }
            }
        }
    }

    @GetMapping("/outdevices")
    public Collection<Long> outdevices(){
        return loanRepository.findByVisszavetteIsNullOrderByTakeDate().stream()
                .map(x->x.getDeviceEntity().getId()).collect(Collectors.toSet());
    }

    @PostMapping("/loanNote")
    public String loanNoteEdit(@RequestBody Map<String, String> requestBody, HttpServletResponse response){
        Optional<LoanEntity> loanEntity = loanRepository.findById(Long.parseLong(requestBody.get("num")));
        if (loanEntity.isEmpty()){
            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
            return "Nein";
        }
        loanService.editNote(loanEntity.get(),requestBody.get("resStr"));
        return "Alma";
    }
}

File: src/main/java/com/sem/keeper/web/AjaxController.kt
package com.sem.keeper.web

import com.sem.keeper.service.command.ChangeLoanNoteCommand
import com.sem.keeper.entity.DeviceEntity
import com.sem.keeper.entity.LoanEntity
import com.sem.keeper.entity.LoanRequestEntity
import com.sem.keeper.entity.UserEntity
import com.sem.keeper.repo.DeviceRepository
import com.sem.keeper.repo.LoanRepository
import com.sem.keeper.repo.LoanRequestRepository
import com.sem.keeper.repo.UserRepository
import com.sem.keeper.service.LoanService
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.web.bind.annotation.*
import java.util.*
import java.util.stream.Collectors
import javax.servlet.http.HttpServletResponse
import javax.servlet.http.HttpSession

@Service
@RestController
@RequestMapping("/ajax")
class AjaxController(
    private val devices: DeviceRepository,
    private val loanRepository: LoanRepository,
    private val loanService: LoanService,
    private val userRepository: UserRepository,
    private val loanRequestRepository: LoanRequestRepository
) {
    @GetMapping("/myReqs")
    fun myReqs(@SessionAttribute user: UserEntity): Iterable<LoanRequestEntity> {
        return loanRequestRepository.findByElvinne(user)
    }

    @GetMapping("/devices")
    fun devicePag(@RequestParam page: Int, @RequestParam pageSize: Int): Iterable<DeviceEntity> {
        val pageable: Pageable = PageRequest.of(page, pageSize)
        return devices.findAll(pageable)
    }

    @GetMapping(value = ["/devicesearch", "/devicesearch/{pageNumber}", "/devicesearch/{pageNumber}/{pageSize}"])
    fun searchDevice(
        @RequestParam term: String?,
        @PathVariable pageNumber: Optional<Int>,
        @PathVariable pageSize: Optional<Int>
    ): Page<DeviceEntity?> {
        val page = pageNumber.orElse(0)
        val pageable = PageRequest.of(page!!, pageSize.orElse(10)!!)
        return devices.findByNameContainingIgnoreCase(term, pageable)
    }

    @GetMapping("/usersearch/{pageNumber}/{pageSize}")
    fun searchUser(
        @RequestParam term: String?,
        @PathVariable pageNumber: Optional<Int>,
        @PathVariable pageSize: Optional<Int>
    ): Page<UserEntity> {
        val page = pageNumber.orElse(0)
        val pageable = PageRequest.of(page!!, pageSize.orElse(10)!!)
        return userRepository.findByFullNameContainingIgnoreCase(term!!, pageable)
    }

    @GetMapping(value = ["/loanlist", "/loanlist/{pageNumber}/{pageSize}"])
    fun loanlist(
        session: HttpSession,
        @RequestParam filter: Optional<String>,
        @PathVariable pageNumber: Optional<Int>,
        @PathVariable pageSize: Optional<Int>
    ): Page<LoanEntity> {
        val pageable: Pageable = PageRequest.of(
            pageNumber.orElse(0),
            pageSize.orElse(10)
        )
        return if (filter.isEmpty || filter.get() == "all") {
            loanRepository.findAll(pageable)
        } else {
            when (filter.get()) {
                "mine" -> {
                    val user = session.getAttribute("user") as UserEntity
                    loanRepository.findByElvitte(user, pageable)
                }
                "out" -> {
                    loanRepository.findByVisszavetteIsNullOrderByTakeDate(pageable)
                }
                else -> {
                    loanRepository.findAll(pageable)
                }
            }
        }
    }

    @GetMapping("/outdevices")
    fun outdevices(): Collection<Long?> {
        return loanRepository.findByVisszavetteIsNullOrderByTakeDate().stream()
            .map { (_, deviceEntity) -> deviceEntity!!.id }.collect(Collectors.toSet())
    }

    @PostMapping("/loanNote")
    fun loanNoteEdit(@RequestBody requestBody: Map<String, String>, response: HttpServletResponse): String {
        val loanEntity = loanRepository.findById(requestBody["num"]!!.toLong())
        if (loanEntity.isEmpty) {
            response.status = HttpServletResponse.SC_BAD_REQUEST
            return "Nein"
        }
        requestBody["resStr"]?.let {
            loanService.editNote(ChangeLoanNoteCommand(loanEntity.get(), it))
            return "Alma"
        }
        return "Nein"
    }
}
